ggml/src/ggml-amx.cpp

```diff
@@ -0,0 +1,461 @@
+  return true;
+#elif defined(_WIN32)
+  return true;
+#endif
+}
+
+ggml_backend_t ggml_backend_amx_init() {
+
+  // invoke a Linux system call to request access to AMX features
+  ggml_amx_init();
+
+  // backend context
+  ggml_backend_amx_context * ctx = new ggml_backend_amx_context;
+
+  // ggml amx backend
+  ggml_backend_t backend = new ggml_backend {
+    /* .guid      = */ ggml_backend_amx_guid(),
+    /* .interface = */ ggml_backend_amx_i,
+    /* .device    = */ ggml_backend_reg_dev_get(ggml_backend_amx_reg(), 0),
+    /* .context   = */ ctx,
+  };
+
+  return backend;
+}
+
+bool ggml_backend_is_amx(ggml_backend_t backend) {
+  return backend != NULL && ggml_guid_matches(backend->guid, ggml_backend_amx_guid());
+}
+
+void ggml_backend_amx_set_n_threads(ggml_backend_t backend_amx, int n_threads) {
+  GGML_ASSERT(ggml_backend_is_amx(backend_amx));
+
+  ggml_backend_amx_context * ctx = (ggml_backend_amx_context *)backend_amx->context;
+  ctx->n_threads = n_threads;
+}
+
+// device interface
+
+static const char * ggml_backend_amx_device_get_name(ggml_backend_dev_t dev) {
+  return "AMX";
+
+  GGML_UNUSED(dev);
+}
+
+static const char * ggml_backend_amx_device_get_description(ggml_backend_dev_t dev) {
+  return "Intel Advanced Matrix Extensions";
+
+  GGML_UNUSED(dev);
+}
+
+static void ggml_backend_amx_device_get_memory(ggml_backend_dev_t dev, size_t * free, size_t * total) {
+  // TODO
+  *free = 0;
+  *total = 0;
+
+  GGML_UNUSED(dev);
+}
+
+static enum ggml_backend_dev_type ggml_backend_amx_device_get_type(ggml_backend_dev_t dev) {
+  return GGML_BACKEND_DEVICE_TYPE_CPU;
+
+  GGML_UNUSED(dev);
+}
+
+static void ggml_backend_amx_device_get_props(ggml_backend_dev_t dev, struct ggml_backend_dev_props * props) {
+  props->name        = ggml_backend_amx_device_get_name(dev);
+  props->description = ggml_backend_amx_device_get_description(dev);
+  props->type        = ggml_backend_amx_device_get_type(dev);
+  ggml_backend_amx_device_get_memory(dev, &props->memory_free, &props->memory_total);
+  props->caps = {
+    /* .async                 = */ false,
+    /* .host_buffer           = */ false,
+    /* .buffer_from_host_ptr  = */ true,
+    /* .events                = */ false,
+  };
+}
+
+static ggml_backend_t ggml_backend_amx_device_init(ggml_backend_dev_t dev, const char * params) {
+  return ggml_backend_amx_init();
+
+  GGML_UNUSED(dev);
+  GGML_UNUSED(params);
+}
+
+static ggml_backend_buffer_type_t ggml_backend_amx_device_get_buffer_type(ggml_backend_dev_t dev) {
+  return ggml_backend_amx_buffer_type();
+
+  GGML_UNUSED(dev);
+}
+
+static ggml_backend_buffer_t ggml_backend_amx_device_buffer_from_ptr(ggml_backend_dev_t dev, void * ptr, size_t size, size_t max_tensor_size) {
+  GGML_ASSERT((uintptr_t)ptr % TENSOR_ALIGNMENT == 0 && "buffer pointer must be aligned");
+  return ggml_backend_buffer_init(ggml_backend_amx_buffer_type(), ggml_backend_amx_buffer_interface, ptr, size);
+
+  GGML_UNUSED(dev);
+  GGML_UNUSED(max_tensor_size);
+}
```

Comment:
I think that this backend should not implement this function since it modifies the layout of the tensors. This is mainly intended to be used with mmap, which means that the backend must be able to use the tensors in the standard ggml layout.

Reply:
Oh thanks for the headsup!